### PR TITLE
Initial work to use base client

### PR DESCRIPTION
**Pull request by Marenz**
Currently on hold as the base client does not seem to easily support using a
custom root certificate authority.
At this point we still need to use that as we're still waiting on [proper certificates](https://github.com/frequenz-io/infrastructure/issues/73)

So this is blocked by either:
* BaseClient supporting custom root certificates.
* Infra giving us an officially recognized SSL certificate.

### CODE REVIEW

**Comment by llucax** [comment]
> BaseClient supporting custom root certificates

Can you create an issue (or better a PR :innocent: ) with what's needed? An example of how you are creating the grpc channel would help.

**Comment by Marenz** [comment]
You can see the according code here: https://github.com/frequenz-floss/frequenz-client-dispatch-python/blob/v0.x.x/src/frequenz/client/dispatch/__main__.py#L57
We add a repo-local certificate here as root so that our client accepts the self-signed certificate of the service.

**Comment by llucax** [comment]
So looking at that now, and you seem to have a function that accepts more than just the root cert, so for now I will only implement passing a root cert to the base client. Let me know if I am misreading this and you also need to pass a private key and a cert chain.

**Comment by llucax** [comment]
Also, any reason for storing the root cert in the repo and using `files(...)`? Shouldn't the root cert be provided by whoever actually instantiates the client and not the client package itself?

**Comment by llucax** [comment]
In any case, the base client will take the root cert as a path to a file and let clients decide how to do the config, but it seems weird to ship a root cert and force any client users to use that root cert.

**Comment by Marenz** [comment]
> Shouldn't the root cert be provided by whoever actually instantiates the client and not the client package itself?

no, because they wouldn't have the cert ;) 

We generated the cert, we signed the server certs with it and we also ship it in the client because that's the only way they can validate the server cert.

**Comment by llucax** [comment]
> We generated the cert, we signed the server certs with it and we also ship it in the client because that's the only way they can validate the server cert.

You can publish the certs somewhere, like you do with you gpg public key, no? Anyway:

- https://github.com/frequenz-floss/frequenz-client-base-python/pull/73

**Comment by llucax** [comment]
I might need to understand the cert part better, as if the cert really needs to be shipped by the client then passing it via URL doesn't make a lot of sense...

**Comment by Marenz** [comment]
> You can publish the certs somewhere, like you do with you gpg public key, no? Anyway:

I mean sure, but why give our client users so much work if we can just ship it straight with the client?
It's the only place were its used and useful. 
Remember also that this is a temporary solution until infrastructure gives me a proper certificate signed by a real CA... https://github.com/frequenz-io/infrastructure/issues/73

**Comment by llucax** [comment]
Yeah, ok, given let's encrypt I hope we can get normal certificates without needing a root cert. Not sure if we should event merge the PR, maybe for "advanced use" the client should just accept a pre created channel instead of a URL.

File: src/frequenz/client/dispatch/_client.py [hunk 1]
```diff
@@ -37,23 +40,43 @@
...
+            svc_addr,
+            dispatch_pb2_grpc.MicrogridDispatchServiceStub,
+            connect=connect,
+            channel_defaults=ChannelOptions(
+                port=DEFAULT_DISPATCH_PORT,
+                ssl=SslOptions(
+                    enabled=True,
+                    root_certificates=Path(
+                        str(
+                            files("frequenz.client.dispatch").joinpath("certs/root.crt")
```
Comment (llucax) [inline src/frequenz/client/dispatch/_client.py:73]: Doesn't this already return a `Path`?

Comment (Marenz) [reply src/frequenz/client/dispatch/_client.py:73]: Unfortunately, `joinpath` seems to return a `Traversable`

Comment (llucax) [reply src/frequenz/client/dispatch/_client.py:73]:
```py
>>> import pathlib
>>> p = pathlib.Path("/tmp").joinpath("lala/papap")
>>> p
PosixPath('/tmp/lala/papap')
>>> isinstance(p, pathlib.Path)
True
```
:thinking:

Comment (llucax) [reply src/frequenz/client/dispatch/_client.py:73]: And:
```py
>>> p2 = importlib.resources.files("frequenz.sdk")
>>> p2
PosixPath('/home/luca/devel/sdk/src/frequenz/sdk')
>>> isinstance(p2, pathlib.Path)
True
```

So, a `Path` is probably a `Traversable`, but it is still a `Path` :D

Comment (llucax) [reply src/frequenz/client/dispatch/_client.py:73]: Oh, but if it is typed as `Traversable` you would have to cast it I guess. Nevermind, all good, nothing to see here.

File: src/frequenz/client/dispatch/_client.py [hunk 2]
```diff
@@ -37,23 +40,43 @@
...
 
-class Client:
+class Client(BaseApiClient[dispatch_pb2_grpc.MicrogridDispatchServiceStub]):
     """Dispatch API client."""
 
     def __init__(
-        self, *, grpc_channel: grpc.aio.Channel, svc_addr: str, key: str
+        self,
+        *,
+        svc_addr: str,
```
Comment (llucax) [inline src/frequenz/client/dispatch/_client.py]: Sorry I missed this before, but could you use `service_url` instead, like the base client, so we have a consistent interface across all the different API clients?
```suggestion
        server_url: str,
```

**Review by llucax — COMMENTED** [review]
Sorry, last change request (don't forget to update release notes too). :innocent:

File: RELEASE_NOTES.md [hunk 3]
```diff
@@ -2,13 +2,16 @@
 
 ## Summary
 
-This release adds support for pagination in the dispatch list request.
+This release includes a new feature for pagination support in the dispatch list request as well as usage of the base-client for setting up the channel and client configuration.
 
 ## Upgrading
 
 - The `Client.list()` function now yields a `list[Dispatch]` representing one page of dispatches
+- `Client.__init__` no longer accepts a `grpc_channel` argument, instead a `svc_addr` argument is required.
```
Comment (llucax) [inline RELEASE_NOTES.md]:
```suggestion
- `Client.__init__` no longer accepts a `grpc_channel` argument, instead a `server_url` argument is required.
```
image

Comment (Marenz) [reply RELEASE_NOTES.md]: that's what you meant :laughing: 
I updated the release notes while rebasing ;)-

**Review by llucax — APPROVED** [review]
:tada:

**Pull request merged**